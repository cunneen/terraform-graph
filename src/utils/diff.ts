import { IDiff, IResource } from '../models'
import isEqual = require('lodash.isequal')

export function getDiff(resource: IResource): IDiff {
  const diff: IDiff = { address: resource.address, action: resource.change.actions.join('-'), changes: [] }
  resource.change.before = resource.change.before || {}
  resource.change.after = resource.change.after || {}
  for (const key of Object.keys(resource.change.after_unknown)) {
    resource.change.after[key] = '(known after apply)'
  }
  for (const key of getKeys(resource.change.before, resource.change.after)) {
    const action = getAction(resource.change.before[key], resource.change.after[key])
    if (action === 'no-op') continue
    diff.changes.push({
      key,
      action,
      before: getContent(resource.change.before[key]),
      after: getContent(resource.change.after[key]),
    })
  }
  return diff
}

function getKeys(before: { [key: string]: unknown }, after: { [key: string]: unknown }) {
  const beforeKeys = Object.keys(before)
  const afterKeys = Object.keys(after)

  const keys = new Set<string>()
  for (let i = 0; i < Math.max(beforeKeys.length, afterKeys.length); i++) {
    if (beforeKeys[i]) keys.add(beforeKeys[i])
    if (afterKeys[i]) keys.add(afterKeys[i])
  }

  return [...keys]
}

function getAction(before: unknown, after: unknown) {
  if (isEqual(before, after)) return 'no-op'
  if (!before) return 'create'
  if (!after) return 'delete'
  return 'update'
}

function getContent(value: unknown) {
  if (typeof value === 'object') {
    return JSON.stringify(value, null, 2)
  }
  if (typeof value === 'string') {
    try {
      return getContent(JSON.parse(value))
    } catch (error) {
      return value || '(none)'
    }
  }
  return String(value ?? '(none)')
}
