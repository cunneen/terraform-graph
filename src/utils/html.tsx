import { Fragment } from 'react'
import { ITree } from '../models'

export function renderHtml(tree: ITree, scriptUrl: string, stylesUrl: string[], cspSource: string): JSX.Element {
  return (
    <html lang="en">
      <head>
        <meta charSet="UTF-8" />
        <meta httpEquiv="Content-Security-Policy" content={`default-src ${cspSource}`} />
        <meta name="viewport" content="width=device-width, initial-scale=1.0" />
        <title>Terraform graph</title>

        {stylesUrl.map((styleUrl) => (
          <link href={styleUrl} rel="stylesheet" />
        ))}
      </head>
      <body>
        <div id="resources">
          <ul>{renderTree(tree)}</ul>
        </div>
        <div id="properties">{renderTables(tree)}</div>
        <script src={scriptUrl}></script>
      </body>
    </html>
  )
}

function renderTree(tree: ITree): JSX.Element {
  const children = Object.values(tree.children)
  return (
    <li>
      {tree.diff ? (
        <label htmlFor={encodeURIComponent(tree.diff.address)}>
          {renderActions(tree.diff.actions)} {tree.label}
        </label>
      ) : (
        <span>{tree.label}</span>
      )}
      {Boolean(children.length) && (
        <ul>
          {children.map((child, key) => (
            <Fragment key={key}>{renderTree(child)}</Fragment>
          ))}
        </ul>
      )}
    </li>
  )
}

function renderTables(tree: ITree): JSX.Element {
  const children = Object.values(tree.children)
  return (
    <>
      {tree.diff && (
        <>
          <input id={encodeURIComponent(tree.diff.address)} type="radio" name="table" />
          <table>
            <caption className="address">
              {tree.diff.address}
              <div className="monaco-toolbar">
                <div className="monaco-action-bar">
                  <ul className="actions-container" role="toolbar">
                    <li className="action-item" role="presentation" custom-hover="true">
                      <span
                        className="action-label codicon codicon-copy"
                        role="button"
                        aria-label="Copy resource address"
                        data-content={tree.diff.address}
                      ></span>
                    </li>
                  </ul>
                </div>
              </div>
            </caption>
            <tbody>
              {tree.diff.changes.map((change) => (
                <tr key={change.key}>
                  <td className="nowrap">
                    {renderActions([change.action])} {change.key}
                  </td>
                  <td>
                    <pre>{change.before}</pre>
                  </td>
                  <td>â†’</td>
                  <td>
                    <pre>{change.after}</pre>
                  </td>
                </tr>
              ))}
            </tbody>
          </table>
        </>
      )}
      {Boolean(children.length) && children.map((child, key) => <Fragment key={key}>{renderTables(child)}</Fragment>)}
    </>
  )
}

function renderActions(actions: string[]) {
  const actionSymbol: Record<string, string> = {
    create: '+',
    update: '~',
    delete: '-',
  }

  return actions.map((action) => (
    <span key={action} className={action}>
      {actionSymbol[action]}
    </span>
  ))
}
