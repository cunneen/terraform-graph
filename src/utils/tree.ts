import { IPlan, ITree } from '../models'
import { getDiff } from './diff'

export function generateTree(plan: IPlan): ITree {
  const tree: ITree = { label: 'root', children: {} }

  for (const resource of plan.resource_changes) {
    if (resource.change.actions[0] === 'no-op') continue
    let curr = tree
    for (const label of splitAddress(resource.address)) {
      if (!curr.children[label]) {
        curr.children[label] = { label, children: {} }
      }
      curr = curr.children[label]
    }
    curr.diff = getDiff(resource)
  }

  return tree
}

function splitAddress(address: string) {
  const labels: string[] = []
  let label = ''
  let inForEach = false

  for (const char of address) {
    if (char === '[') inForEach = true
    else if (char === ']') inForEach = false

    if (char === '[' || char === ']' || (char === '.' && !inForEach)) {
      labels.push(label)
      label = ''
    } else if (char !== '"' && char !== "'") {
      label += char
    }
  }

  labels.push(label)

  return labels.filter(Boolean)
}
