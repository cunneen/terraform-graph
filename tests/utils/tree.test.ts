import { generateTree } from '../../src/utils/tree'
import { mockPlan, mockResource, mockResourceChange } from '../mocks'

describe('generateTree', () => {
  it('should return empty tree', () => {
    const plan = mockPlan({ resource_changes: [] })
    const tree = generateTree(plan)
    expect(tree).toMatchSnapshot()
  })

  it('should not add no-op change in tree', () => {
    const plan = mockPlan({ resource_changes: [mockResource({ change: mockResourceChange({ actions: ['no-op'] }) })] })
    const tree = generateTree(plan)
    expect(tree).toMatchSnapshot()
  })

  it('should add change in tree', () => {
    const plan = mockPlan()
    const tree = generateTree(plan)
    expect(tree).toMatchSnapshot()
  })
})
