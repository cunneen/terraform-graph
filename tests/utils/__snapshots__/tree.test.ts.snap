// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateTree > should add change in tree 1`] = `
{
  "children": {
    "resource": {
      "children": {
        "address": {
          "children": {},
          "diff": {
            "actions": [],
            "address": "resource.address",
            "changes": [],
          },
          "label": "address",
        },
      },
      "label": "resource",
    },
  },
  "label": "root",
}
`;

exports[`generateTree > should not add no-op change in tree 1`] = `
{
  "children": {},
  "label": "root",
}
`;

exports[`generateTree > should return empty tree 1`] = `
{
  "children": {},
  "label": "root",
}
`;
